include/pybind11/detail/internals.h

```diff
@@ -281,20 +282,59 @@ inline internals **&get_internals_pp() {
 }
 
 inline void translate_exception(std::exception_ptr p) {
+    if (!p) {
+        return;
+    }
+    auto raise_err = PyErr_SetString;
+#if PY_VERSION_HEX >= 0x03030000
+    // handles nested C++ exceptions if supported
     try {
-        if (p) std::rethrow_exception(p);
-    } catch (error_already_set &e)           { e.restore();                                    return;
-    } catch (const builtin_exception &e)     { e.set_error();                                  return;
-    } catch (const std::bad_alloc &e)        { PyErr_SetString(PyExc_MemoryError,   e.what()); return;
-    } catch (const std::domain_error &e)     { PyErr_SetString(PyExc_ValueError,    e.what()); return;
-    } catch (const std::invalid_argument &e) { PyErr_SetString(PyExc_ValueError,    e.what()); return;
-    } catch (const std::length_error &e)     { PyErr_SetString(PyExc_ValueError,    e.what()); return;
-    } catch (const std::out_of_range &e)     { PyErr_SetString(PyExc_IndexError,    e.what()); return;
-    } catch (const std::range_error &e)      { PyErr_SetString(PyExc_ValueError,    e.what()); return;
-    } catch (const std::overflow_error &e)   { PyErr_SetString(PyExc_OverflowError, e.what()); return;
-    } catch (const std::exception &e)        { PyErr_SetString(PyExc_RuntimeError,  e.what()); return;
+        // Rethrow the exception_ptr to recove type info.
+        // Dereferencing an exception_ptr is UB otherwise.
+        std::rethrow_exception(p);
+    } catch (const std::nested_exception &nep) {
+        std::exception_ptr nested = nep.nested_ptr();
+        if (nested != nullptr && nested != p) {
+            translate_exception(nested);
+            raise_err = raise_from;
+        }
+    } catch (...) {
+    }
+#endif
+    try {
+        std::rethrow_exception(p);
+    } catch (error_already_set &e) {
+        e.restore();
+        return;
```

Comment:
Are the many `return;` needed (to make some compiler(s) happy)?
Would be nice to get rid of them. Or add a comment (maybe) right before the `try`, something like:

// The returns inside the catch blocks are needed to silence compiler warnings.
(If that's true.)

Reply:
I don't know why the return's are here. There were here before the diff.